phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java

```diff
@@ -3416,6 +3437,167 @@ private MetaDataMutationResult mutateColumn(
+                    // hbase.rpc.timeout. Even if the future times out, this runnable can be in
+                    // RUNNING state and will not be interrupted.
+                    service.invalidateServerMetadataCache(controller, request);
+                    LOGGER.info("Invalidating metadata cache for tenantID: {}, tableName: {}"
+                            + " on region server: {} completed successfully and it took {} ms",
+                            tenantIDStr, fullTableName, serverName,
+                            innerWatch.stop().elapsedMillis());
+                    // TODO Create a histogram metric for time taken for invalidating the cache.
+                } catch (ServiceException se) {
+                    LOGGER.error("Invalidating metadata cache for tenantID: {}, tableName: {}"
+                                    + " failed for regionserver {}", tenantIDStr, fullTableName,
+                            serverName, se);
+                    IOException ioe = ServerUtil.parseServiceException(se);
+                    throw new CompletionException(ioe);
+                }
+            });
+            futures.add(future);
+            map.put(future, serverName);
+        }
+
+        // Here we create one master like future which tracks individual future
+        // for each region server.
+        CompletableFuture<Void> allFutures = CompletableFuture.allOf(
+                futures.toArray(new CompletableFuture[0]));
+        try {
+            allFutures.get(metadataCacheInvalidationTimeoutMs, TimeUnit.MILLISECONDS);
+        } catch (Throwable t) {
+            List<ServerName> failedServers = getFailedServers(futures, map);
+            LOGGER.error("Invalidating metadata cache for tenantID: {}, tableName: {} failed for "
+                    + "region servers: {}", tenantIDStr, fullTableName, failedServers, t);
+            if (isRetry) {
+                // If this is a retry attempt then just fail the operation.
+                if (allFutures.isCompletedExceptionally()) {
+                    if (t instanceof ExecutionException) {
+                        t = t.getCause();
+                    }
+                }
+                throw t;
+            } else {
+                // This is the first attempt, we can retry once.
+                // Indicate that this is a retry attempt.
+                invalidateServerMetadataCacheWithRetries(admin, failedServers,
+                        tenantId, schemaName, tableOrViewName, true);
+            }
+        }
+    }
+
+    /*
+        Get the list of regionservers that failed the invalidateCache rpc.
+     */
+    private List<ServerName> getFailedServers(List<CompletableFuture<Void>> futures,
+                                              Map<Future, ServerName> map) {
+        List<ServerName> failedServers = new ArrayList<>();
+        for (CompletableFuture completedFuture : futures) {
+            if (completedFuture.isDone() == false) {
```

Comment:
I am not familiar with futures, but what happens in case of a timeout? Will it have a `false` value? In the timeout case, won't the remote RPC thread be in a stuck state? What is the chance of a retry making progress, besides initiating a concurrent request?

Reply:
> Will it have a false value?

Yes, it will have false value. We also have [test case](https://github.com/apache/phoenix/blob/PHOENIX-6883-feature/phoenix-core/src/it/java/org/apache/phoenix/end2end/InvalidateMetadataCacheIT.java#L67) for this.

> In the timeout case, won't the remote RPC thread be in a stuck state? What is the chance of a retry making progress, besides initiating a concurrent request?

In a later jira [PHOENIX-7115](https://issues.apache.org/jira/browse/PHOENIX-7115), we have added special handler threads on each regionserver just to handle invalidate cache RPCs. Earlier invalidate cache RPCs were handled by default handler threads. So the probability of these RPC threads being stuck is very small but if it happens we can increase the number of threads if necessary.